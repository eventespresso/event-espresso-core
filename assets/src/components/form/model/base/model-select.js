/**
 * External imports
 */
import Select from 'react-select';
import { Component, Fragment } from '@wordpress/element';
import {
	isEmpty,
	uniqueId,
	find,
	isUndefined,
	isFunction,
	isMap,
} from 'lodash';
import PropTypes from 'prop-types';
import isShallowEqual from '@wordpress/is-shallow-equal';

/**
 * WP dependencies
 */
import { withSelect } from '@wordpress/data';

/**
 * Internal imports
 */
import buildOptions from './build-options';
import { MODEL_NAMES } from '../../../../data/model/index';
import {
	REACT_SELECT_DEFAULTS,
	REACT_SELECT_TYPES,
} from './default-select-configuration';
import {
	ALLOWED_ORDER_VALUES,
	QUERY_ORDER_DESC,
} from '../../../../data/model/base';

/**
 * ModelSelect component.
 * This is a component that will generate a react-select input for a given
 * model and its entities (provided via props).
 *
 * @see https://deploy-preview-2289--react-select.netlify.com/props#prop-types
 *   for options that can be passed through via the selectConfiguration prop.
 *
 * @param { Object } selectConfiguration 	An object containing options for the
 *                                          react-select component.
 * @param { Array } modelEntities          	Array of model entities
 * @param { string } modelName              The name of the Model the entities
 *                                          belong to.
 * @param {Function} mapOptionsCallback  	This function will receive by default
 * 											the modelEntities and optionsEntityMap,
 * 											and is expected to return an array of options
 * 											to be used for the react-select component.
 * @param { Object } optionsEntityMap    	Object of key value pairs where values are the
 *                                        	'label:value' pairings used by `mapOptionsCallback`
 * @param { string } mapSelection        	Determines which optionEntityMap pairing to use
 */
export class ModelSelect extends Component {
	state = {};

	static propTypes = {
		selectConfiguration: PropTypes.shape({
			...REACT_SELECT_TYPES,
		}),
		modelEntities: PropTypes.array,
		modelName: PropTypes.oneOf(MODEL_NAMES).isRequired,
		mapOptionsCallback: PropTypes.func,
		optionsEntityMap: PropTypes.object,
		mapSelection: PropTypes.string,
		queryData: PropTypes.shape({
			limit: PropTypes.number,
			orderBy: PropTypes.string,
			order: PropTypes.oneOf(ALLOWED_ORDER_VALUES),
		}),
		getQueryString: PropTypes.func,
		label: PropTypes.string,
	};

	static defaultProps = {
		selectConfiguration: {
			...REACT_SELECT_DEFAULTS,
			name: uniqueId('model-select-'),
		},
		modelEntities: [],
		mapOptionsCallback: buildOptions,
		mapSelection: 'default',
		queryData: {
			limit: 100,
			order: QUERY_ORDER_DESC,
		},
		label: '',
		getQueryString: () => '',
	};

	constructor(props) {
		super(props);
		this.state = ModelSelect.setStateFromProps(props);
	}

	/**
	 * Sets the state from provided props
	 *
	 * @param {Object} props
	 * @return {{
	 * isClearable,
	 * isLoading,
	 * placeholder,
	 * options: (*|Array),
	 * value: *
	 * }}
	 * Object to replace state.
	 */
	static setStateFromProps(props) {
		const { selectConfiguration } = props;
		const options = ModelSelect.getOptions(props);
		const selectedValue = ModelSelect.getOptionObjectForValue(
			selectConfiguration.defaultValue,
			options
		);
		const updated = {
			options,
			value: selectedValue,
		};

		return {
			...REACT_SELECT_DEFAULTS,
			...selectConfiguration,
			...updated,
		};
	}

	/**
	 * Sets up options for the select control from the incoming props.
	 *
	 * @param {Object} props
	 * @return {Array<Object>} Array of options for select control
	 */
	static getOptions(props) {
		const {
			mapOptionsCallback,
			modelEntities,
			optionsEntityMap,
			mapSelection,
		} = props;
		if (isFunction(mapOptionsCallback) && !isEmpty(modelEntities)) {
			return mapOptionsCallback(
				modelEntities,
				optionsEntityMap,
				mapSelection
			);
		}
		return [];
	}

	/**
	 * Given a value, returns the corresponding option object.
	 *
	 * @param {*} value
	 * @param {Array<Object>} options
	 * @return {Object|null} The option object for the given value or null.
	 */
	static getOptionObjectForValue(value, options) {
		if (!isEmpty(options)) {
			const match = find(options, function (option) {
				return option.value === value;
			});
			return !isUndefined(match) ? match : null;
		}
		return null;
	}

	/**
	 * Helper method for determining whether state should be updated.
	 * Criteria is whether the keys from the incoming entities are equivalent
	 * (equivalency means in same order as well)
	 *
	 * @param {Object} prevProps
	 * @param {Object} nextProps
	 * @return {boolean} True means state should be updated.
	 */
	static shouldUpdateState(prevProps, nextProps) {
		if (prevProps === nextProps) {
			return false;
		}

		const { selectConfiguration: prevConfiguration } = prevProps;
		const { selectConfiguration: nextConfiguration } = nextProps;

		// if defaultValue has changed (selected value) then update state
		if (prevConfiguration.defaultValue !== nextConfiguration.defaultValue) {
			return true;
		}

		if (prevConfiguration.isLoading !== nextConfiguration.isLoading) {
			return true;
		}

		// shallow compare of keys but only if we can
		if (
			!isMap(prevProps.modelEntities) ||
			!isMap(nextProps.modelEntities)
		) {
			return true;
		}

		if (
			!isShallowEqual(
				Array.from(prevProps.modelEntities.keys()),
				Array.from(nextProps.modelEntities.keys())
			)
		) {
			return true;
		}
	}

	componentDidMount() {
		this.setState(ModelSelect.setStateFromProps(this.props));
	}

	componentDidUpdate(prevProps) {
		if (ModelSelect.shouldUpdateState(prevProps, this.props)) {
			this.setState(ModelSelect.setStateFromProps(this.props));
		}
	}

	/**
	 * Returns the label for the select control
	 *
	 * @return {string} The label to use.
	 */
	getSelectLabel() {
		const { label, selectConfiguration } = this.props;
		return label ? (
			<label htmlFor={selectConfiguration.name}>{label}</label>
		) : (
			''
		);
	}

	render() {
		return (
			<Fragment>
				{this.getSelectLabel()}
				<Select {...this.state} />
			</Fragment>
		);
	}
}

/**
 * The ModelSelect Component wrapped in the `withSelect` higher order component.
 * This subscribes the ModelSelect component to the state maintained via the
 * eventespresso/lists store.
 */
export default withSelect((select, ownProps) => {
	const { getQueryString, modelName, selectConfiguration } = ownProps;
	const queryString = getQueryString(ownProps.queryData);
	const { getEntities, isRequestingEntities } = select('eventespresso/lists');
	return {
		...ModelSelect.defaultProps,
		...ownProps,
		modelEntities: getEntities(modelName, queryString),
		selectConfiguration: {
			...ModelSelect.defaultProps.selectConfiguration,
			...selectConfiguration,
			isLoading: isRequestingEntities(modelName, queryString),
		},
	};
})(ModelSelect);
