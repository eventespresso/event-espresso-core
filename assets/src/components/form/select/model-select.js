/**
 * External imports
 */
import Select from 'react-select';
import { Component, Fragment } from '@wordpress/element';
import { isEmpty, uniqueId, find, isUndefined, isFunction } from 'lodash';
import PropTypes from 'prop-types';

/**
 * WP dependencies
 */
import { withSelect } from '@wordpress/data';

/**
 * Internal imports
 */
import buildOptions from './build-options';
import { MODEL_NAMES } from '../../../data/model';
import {
	REACT_SELECT_DEFAULTS,
	REACT_SELECT_TYPES,
} from './default-select-configuration';

/**
 * ModelSelect component.
 * This is a component that will generate a react-select input for a given
 * model and its entities (provided via props).
 *
 * @see https://deploy-preview-2289--react-select.netlify.com/props#prop-types
 *   for options that can be passed through via the selectConfiguration prop.
 *
 * @param { Object } selectConfiguration 	An object containing options for the
 *                                          react-select component.
 * @param { Array } modelEntities          	Array of model entities
 * @param { string } modelName              The name of the Model the entities
 *                                          belong to.
 * @param { function } mapOptionsCallback  	This function will receive by default
 * 											the modelEntities and optionsEntityMap,
 * 											and is expected to return an array of options
 * 											to be used for the react-select component.
 * @param { Object } optionsEntityMap    	Object of key value pairs where values are the
 *                                        	'label:value' pairings used by `mapOptionsCallback`
 * @param { string } mapSelection        	Determines which optionEntityMap pairing to use
 */
export class ModelSelect extends Component {
	state = {};

	static propTypes = {
		selectConfiguration: PropTypes.shape( {
			...REACT_SELECT_TYPES,
		} ),
		modelEntities: PropTypes.array,
		modelName: PropTypes.oneOf( MODEL_NAMES ),
		mapOptionsCallback: PropTypes.func,
		optionsEntityMap: PropTypes.object,
		mapSelection: PropTypes.string,
		queryData: PropTypes.shape( {
			limit: PropTypes.number,
			orderBy: PropTypes.string,
			order: PropTypes.oneOf( [ 'asc', 'desc' ] ),
		} ),
		getQueryString: PropTypes.func,
		label: PropTypes.string,
	};

	static defaultProps = {
		selectConfiguration: {
			...REACT_SELECT_DEFAULTS,
			name: uniqueId( 'model-select-' ),
		},
		modelEntities: [],
		mapOptionsCallback: buildOptions,
		mapSelection: 'default',
		queryData: {
			limit: 100,
			order: 'desc',
		},
		label: '',
		getQueryString: () => '',
	};

	static getDerivedStateFromProps( props ) {
		const { selectConfiguration } = props;
		const options = ModelSelect.getOptions( props );
		const selectedValue = ModelSelect.getOptionObjectForValue(
			selectConfiguration.defaultValue, options
		);
		const updated = {
			options,
			value: selectedValue,
		};

		return {
			...REACT_SELECT_DEFAULTS,
			...selectConfiguration,
			...updated,
		};
	}

	static getOptions( props ) {
		const {
			mapOptionsCallback,
			modelEntities,
			optionsEntityMap,
			mapSelection,
		} = props;
		if ( isFunction( mapOptionsCallback ) && ! isEmpty( modelEntities ) ) {
			return mapOptionsCallback(
				modelEntities,
				optionsEntityMap,
				mapSelection
			);
		}
		return [];
	}

	static getOptionObjectForValue( value, options ) {
		if ( ! isEmpty( options ) ) {
			const match = find( options, function( option ) {
				return option.value === value;
			} );
			return ! isUndefined( match ) ?
				match :
				null;
		}
		return null;
	}

	getSelectLabel() {
		const { label, selectConfiguration } = this.props;
		return label ?
			<label htmlFor={ selectConfiguration.name }>{ label }</label> :
			'';
	}

	render() {
		return (
			<Fragment>
				{ this.getSelectLabel() }
				<Select { ...this.state } />
			</Fragment>
		);
	}
}

/**
 * The ModelSelect Component wrapped in the `withSelect` higher order component.
 * This subscribes the ModelSelect component to the state maintained via the
 * eventespresso/lists store.
 */
export default withSelect( ( select, ownProps ) => {
	const { getQueryString, modelName, selectConfiguration } = ownProps;
	const queryString = getQueryString( ownProps.queryData );
	const { getItems, isRequestingItems } = select( 'eventespresso/lists' );
	return {
		...ModelSelect.defaultProps,
		...ownProps,
		modelEntities: getItems( modelName, queryString ),
		selectConfiguration: {
			...ModelSelect.defaultProps.selectConfiguration,
			...selectConfiguration,
			isLoading: isRequestingItems( modelName, queryString ),
		},
	};
} )( ModelSelect );
