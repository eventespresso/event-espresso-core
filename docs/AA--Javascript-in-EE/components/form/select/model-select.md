# `<ModelSelect />` and `<ModelEnhancedSelect />`

ModelSelect is a component that exposed the [`react-select`](https://deploy-preview-2289--react-select.netlify.com/home) component for displaying provided EE model data for selection. While it can be used on its own, its intended to be utilized within more concrete "wrapper" components specific to a model (eg `EventSelect`) that are generated using [`createModelSelect`](./create-model-select.md).  There are actually 2 components exposed for `ModelSelect`:  a standalone component, `<ModelSelect />` and a `withSelect` Higher Order Component wrapped default export which is exposed as `<ModelEnhancedSelect />`

The HOC automatically takes care of populating the select options as directed from provided props and via subscription to the related `@eventespresso/lists` store.

## `ModelSelect` Props

### selectConfiguration

This is an object that can contain any values set by components higher in the tree that you want to pass down to the `react-select` component.  You can read about all the possible values that can be passed down via [this document](https://deploy-preview-2289--react-select.netlify.com/props).

The following options in this object are generated and set in the ModelSelect component:

| Prop | Purpose |
|------|---------|
`name` | This is automatically set to be a unique name for the select component. It can be overridden.
`options` | This is derived from the incoming array of EE model entity objects.
`value` | This is derived from the incoming `defaultValue` value on the incoming `selectConfiguration` prop.


- Type: `Object`
- Required: No
- Default:
```js
{
    name: uniqueId( 'model-select-' ),
    isClearable: true,
    isLoading: true,
    placeholder: 'Select...'
}
```

### modelEntities

This is expected to be an `Array` of model entity objects for the model corresponding to the value in the `modelName` prop.  

- Type: `Array`
- Required: No
- Default: `[]`

### modelName

This informs what model the incoming modelEntities are for.  This should correspond to the canonical modelName for models as respresented in the EE Rest schema (i.e. Events are `event`, Datetimes are `datetime`).

- Type: `String`
- Required: No
- Default: `''`

### mapOptionsCallback

This should be a function that converts incoming modelEntities into an array of option objects usable by the react-select component.  By default, this is the built-in [`buildOptions`](../../../../../assets/src/components/form/model/base/build-options.js) callback and that can be used as a reference for how any custom function should behave.

- Type: `Function`
- Required: No
- Default: `buildOptions`

### optionsEntityMap:

If null, this is ignored and the default map found in [build-options.js](../../../../../src/components/form/select/build-options.js) is used.  When provided, it is expected to be an object that helps inform how the `mapOptionsCallback` should map entity objects to the `value`, and `label` pairs in the final array of option objects. The map should be shaped like this:

```js
const map = {
    $modelName: {
        label: 'model-field-to-use-for-label'
        value: 'model-field-to-use-for-value'
    }
}
```

**Example:**
```js
const map = {
    event: {
        label: 'EVT_name',
        value: 'EVT_ID',
    }
};

```

- Type: `Object`
- Required: No
- Default: `null`

### label

If provided, a `<label>` will be prepended before the Select element using the text provided by the value on this property.  It will be linked (via the label `for` attribute) to the select input via the `selectConfiguration.name` property.

- Type: `String`
- Required: No
- Default: `''`


## `ModelEnhancedSelect` Props

Along with receiving any props noted for `ModelSelect`.  The enhanced component should also receive the following props:

### modelName

This informs what model the REST request should be done against.  This should correspond to the canonical modelName for models as represented in the EE Rest schema (i.e. Events are `event`, Datetimes are `datetime`).

- Type: `String`
- Required: No
- Default: `''`

### queryData

This is only used by the `ModelEnhancedSelect` HOC (but is passed through the main component).  `queryData` would contain any information required for REST request retrieving entities for the named model (described via `modelName`).

- Type: `Object`
- Required: No
- Default: `{ limit: 100, order: 'desc' }`

### getQueryString

This is another item only used by the `ModelEnhancedSelect` HOC (but is passed through the main component).  `getQueryString` should be a function the returns a query string to append to the REST request endpoint for the given model (described via `modelName`).

- Type: `Function`
- Required: No
- Default: `() => ''`


## Usage

Note: if you plan on exposing something wrapping this component as a component itself on `eejs.components` (this will be the case if you are adding your new component to the `components` folder), then import these components directly via relative path.

These components can be utilized one of two ways:

**Via `eejs.components` global**

```js
const { ModelSelect, ModelEnhancedSelect } = eejs.components;
```

**Via `@eventespresso/components` package**

```js
import { ModelSelect, ModelEnhancedSelect } from '@eventespresso/components'
```

**Note**: this package isn't published yet so you'll need to point to it via your build process. For example, if you use webpack you can register this as an external via:

```js
module.exports = {
    externals: {
        '@eventespresso/components': 'eejs.components',
    }
}
```

You will need to ensure that the event espresso components bundle is a dependency for your implemented code.  You could do something like this:

```php
use EventEspresso\core\domain\services\assets\CoreAssetManager;

wp_register_script(
    'my-script',
    'https://url-to-my-script.com'
    array( CoreAssetManager::JS_HANDLE_COMPONENTS ),
    $version_string,
    true
);
```
