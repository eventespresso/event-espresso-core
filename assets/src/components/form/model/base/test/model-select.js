import { shallow } from 'enzyme';
import { ModelSelect } from '../model-select';

const optionsEntityMap = {
	default: {
		value: 'EVT_ID',
		label: 'EVT_name',
	},
};
const simulatedResponse = [
	{ EVT_ID: 1, EVT_name: 'Event A' },
	{ EVT_ID: 2, EVT_name: 'Event B' },
];

describe('ModelSelect Snapshot with default options (with required modelName)', () => {
	it('should render and match snapshot', () => {
		const wrapper = shallow(<ModelSelect modelName={'event'} />);
		expect(wrapper).toMatchSnapshot();
	});
});

describe('ModelSelect props check', () => {
	it('should render and have passed in props match expectation', () => {
		const wrapper = shallow(
			<ModelSelect
				modelName={'event'}
				modelEntities={simulatedResponse}
				optionsEntityMap={optionsEntityMap}
				selectConfiguration={{
					isClearable: false,
					isLoading: false,
				}}
			/>
		);
		const selectWrapper = wrapper.childAt(0);
		expect(selectWrapper.prop('options')).toEqual([
			{ label: 'Event A', value: 1 },
			{ label: 'Event B', value: 2 },
		]);
		expect(selectWrapper.prop('isClearable')).toEqual(false);
		expect(selectWrapper.prop('isLoading')).toEqual(false);
		expect(selectWrapper.prop('placeholder')).toEqual('Selectâ€¦');
	});
	it('should render a label when selectLabel is provided', () => {
		const wrapper = shallow(
			<ModelSelect
				modelName={'event'}
				modelEntities={simulatedResponse}
				optionsEntityMap={optionsEntityMap}
				label={'Label for Select'}
			/>
		);
		expect(wrapper.childAt(0).text()).toEqual('Label for Select');
	});
});
