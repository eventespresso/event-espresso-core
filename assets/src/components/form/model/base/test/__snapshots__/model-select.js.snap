// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelSelect Snapshot with default options (with required modelName) should render and match snapshot 1`] = `
<Fragment>
  <StateManager
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isClearable={true}
    isLoading={true}
    name="model-select-1"
    options={Array []}
    placeholder="Selectâ€¦"
    value={null}
  />
</Fragment>
`;
